### ISSUE TITLE
tippy bugs whem applying CSP guidelines to prevent XSS attacks

### ISSUE THREAD

**Issue by dalescher** [atomiks/tippyjs#338]
I do not manage to benefit from certain functions when applying CSP guidelines (setting is **display-src 'self'**).

I guess that the issue comes from the fact that this setting forbids the use of inline styling. I manage to get around the issue by using ID and class in the html code and style in the CSS sheet only.

However, I cannot find a way to use tippy arrows with this setting on. If i choose round arrow, the arrow stay within the tooltip, does not rotate, and its size gets bigger and bigger (1st tooltip: small, last tooltip on page: huge). when selecting sharp arrow, it does not even appear.

Is there a way to benefit from all tippy tooltip features while having this CSP setting on?

**Comment by atomiks** [comment]
I don't really know anything about that.

If you want to disallow HTML in tooltip content to prevent XSS you can set `allowHTML: false`.

Otherwise you might have to disable CSP and ensure you don't parse user input as HTML / escape it / validate it before sending it to the user.

**Comment by dalescher** [comment]
Thank you for your answer. I manage to get HTML within the tooltips without having issues with XSS (I just use IDs and class rather than inline html, works fine). The only issue I have is with the arrow that does not show up correctly. I guess it has to do with the fact that arrow code add some inline styling or script that is not allowed with my CSP settings

**Comment by atomiks** [comment]
Can you make a reproducible test case that I can check?

**Comment by atomiks** [comment]
Have you seen: https://stackoverflow.com/questions/40360109/content-security-policy-img-src-self-data

**Comment by dalescher** [comment]
Thank you! It solved the message error I had. However, arrow still appear inside tippy and I still have this console error: 

_Refused to apply inline style because it violates the following Content Security Policy directive: "style-src 'self' https://cdnjs.cloudflare.com https://fonts.googleapis.com". Either the 'unsafe-inline' keyword, a hash ('sha256-nc2qjOk5rtX9UVHwLt6riDajiUVm+k/VWY11CEzcjLQ='), or a nonce ('nonce-...') is required to enable inline execution.
(anonymous) @ utils.js:28_

which point to utils.js file at line 28 =>

`export const injectCSS = css => {
  if (isBrowserSupported) {
    const style = document.createElement('style')
    style.type = 'text/css'
    style.textContent = css
    document.head.insertBefore(style, document.head.firstChild)   // faulty line 28
  }
}
`

it seems that this inject inline code and is then blocked by CSP. I prefer not use 'unsafe-inline' in my CSP policy for style-src. Do you have any idea how to get around this?

**Comment by atomiks** [comment]
Yes, avoid using `all.min.js` (which bundles the CSS), and use the `tippy.min.js` script and `tippy.css` stylesheet by linking them separately.

**Comment by dalescher** [comment]
It works! Thanks a lot atomiks, I would have not succeeded without you.
